demo/js/src/main/scala/games/audio/Data.scala

```diff
@@ -4,24 +4,79 @@ import scala.scalajs.js
 import org.scalajs.dom
 import games.Resource
 import scala.concurrent.Promise
-import games.JsResourceUtil
+import games.JsUtils
 import scala.concurrent.Future
 
+import java.nio.{ ByteBuffer, ByteOrder, FloatBuffer }
+
 import scalajs.concurrent.JSExecutionContext.Implicits.queue
 
-class JsBufferedData private[games] (ctx: JsContext, res: Resource) extends BufferedData {
-  private val decodedDataReady = Promise[js.typedarray.ArrayBuffer]
+class JsRawData private[games] (ctx: WebAudioContext, data: ByteBuffer, format: Format, channels: Int, freq: Int) extends RawData {
+  private val bufferReady = Future {
+    format match {
+      case Format.Float32 => // good to go
+      case _              => throw new RuntimeException("Unsupported data format: " + format)
+    }
+
+    channels match {
+      case 1 => // good to go
+      case 2 => // good to go
+      case _ => throw new RuntimeException("Unsupported channels number: " + channels)
+    }
+
+    val floatBuffer = data.slice().order(ByteOrder.nativeOrder()).asFloatBuffer()
+
+    val sampleCount = floatBuffer.remaining() / channels
+
+    val buffer = ctx.webApi.createBuffer(channels, sampleCount, freq)
+
+    var channelsData = new Array[js.typedarray.Float32Array](channels)
+
+    var channelCur = 0
+    while (channelCur < channels) {
+      channelsData(channelCur) = buffer.getChannelData(channelCur).asInstanceOf[js.typedarray.Float32Array]
+      channelCur += 1
+    }
```

Comment:
You need not use explicit `while` loops in Scala.js. A for loop such as
```scala
for (channelCur <- 0 until channels)
  channelsData(channelCur) = buffer.getChannelData(channelCur).asInstanceOf[js.typedarray.Float32Array]
```
will be optimized away as a `while` loop.